Subject: sysctl for runtime-control of lateral mutex stealing
From: Sven-Thorsten Dietrich <sdietrich@suse.de>

Add /proc/sys/kernel/lateral_steal, to allow switching on and off
equal-priority mutex stealing between threads.

Signed-off-by: Sven-Thorsten Dietrich <sdietrich@suse.de>

---
 kernel/rtmutex.c |    8 ++++++--
 kernel/sysctl.c  |   14 ++++++++++++++
 2 files changed, 20 insertions(+), 2 deletions(-)

Index: linux-2.6.22-SLERT10_BRANCH/kernel/rtmutex.c
===================================================================
--- linux-2.6.22-SLERT10_BRANCH.orig/kernel/rtmutex.c
+++ linux-2.6.22-SLERT10_BRANCH/kernel/rtmutex.c
@@ -25,6 +25,9 @@ int rtlock_timeout __read_mostly = CONFI
 #ifdef CONFIG_ADAPTIVE_RTMUTEX
 int rtmutex_timeout __read_mostly = CONFIG_RTMUTEX_DELAY;
 #endif
+#ifdef CONFIG_RTLOCK_LATERAL_STEAL
+int rtmutex_lateral_steal __read_mostly = 1;
+#endif
 
 /*
  * lock->owner state tracking:
@@ -328,7 +331,8 @@ static inline int lock_is_stealable(stru
 	if (current->prio > pendowner->prio)
 		return 0;
 
-	if (!unfair && (current->prio == pendowner->prio))
+	if (unlikely(current->prio == pendowner->prio) &&
+	   !(unfair && rtmutex_lateral_steal))
 #endif
 		return 0;
 
Index: linux-2.6.22-SLERT10_BRANCH/kernel/sysctl.c
===================================================================
--- linux-2.6.22-SLERT10_BRANCH.orig/kernel/sysctl.c
+++ linux-2.6.22-SLERT10_BRANCH/kernel/sysctl.c
@@ -166,6 +166,10 @@ extern ctl_table inotify_table[];
 int sysctl_legacy_va_layout;
 #endif
 
+#ifdef CONFIG_RTLOCK_LATERAL_STEAL
+extern int rtmutex_lateral_steal;
+#endif
+
 extern int prove_locking;
 extern int lock_stat;
 
@@ -945,6 +949,16 @@ static ctl_table kern_table[] = {
 		.proc_handler   = &proc_dointvec,
        },
 #endif
+#ifdef CONFIG_RTLOCK_LATERAL_STEAL
+	{
+		.ctl_name       = CTL_UNNUMBERED,
+		.procname       = "rtmutex_lateral_steal",
+		.data           = &rtmutex_lateral_steal,
+		.maxlen         = sizeof(int),
+		.mode           = 0644,
+		.proc_handler   = &proc_dointvec,
+       },
+#endif
 #ifdef CONFIG_PROC_FS
 	{
 		.ctl_name       = CTL_UNNUMBERED,
